datafusion/functions/src/datetime/date_part.rs

```diff
@@ -223,9 +239,17 @@ fn seconds(array: &dyn Array, unit: TimeUnit) -> Result<ArrayRef> {
     let secs = as_int32_array(secs.as_ref())?;
     let subsecs = date_part(array, DatePart::Nanosecond)?;
     let subsecs = as_int32_array(subsecs.as_ref())?;
+    // REVIEW: is there a better way to do this? (Is it even a reasonable thing to do?) I want to treat
+    // null as 0, I was expecting some kind of map function (or even better map_if_null) on PrimitiveArray.
+    // Instead I have just discarded the null mask, which feels bad because if I was meant to be able
+    // to do this, I'd expect there to be a function to do it. I think it is also possible that a
+    // null-masked value in the array will not be 0 (I don't think that happens right now, but I think
+    // we're one optimisation away from disaster).
+    let subsecs: PrimitiveArray<Int32Type> =
+        PrimitiveArray::new(subsecs.values().clone(), None);
 
-    let r: Float64Array = binary(secs, subsecs, |secs, subsecs| {
-        (secs as f64 + (subsecs as f64 / 1_000_000_000_f64)) * sf
+    let r: Float64Array = binary(secs, &subsecs, |secs, subsecs| {
+        (secs as f64 + ((subsecs % 1_000_000_000) as f64 / 1_000_000_000_f64)) * sf
```

Comment:
I think the more correct way is to use a kernel other than [binary](https://docs.rs/arrow/latest/arrow/compute/fn.binary.html#details) which s designed to be faster by not having to check each element for null

For example, something like this (untested):
```rust
let r: Float64Array = subsecs.null_count() == 0 {
  // special case binary when there are no nulls for speed
  binary(secs, subsecs, |secs, subsecs| {
        (secs as f64 + (subsecs as f64 / 1_000_000_000_f64)) * sf
   })?
} else {
  // handle nulls
  secs.iter().zip(subsecs.iter()).map(|(secs, subsecs) { 
    secs.map(|secs| {
      let subsecs = subsecs.unwrap_or(0); // default null to 0
        (secs as f64 + ((subsecs % 1_000_000_000) as f64 / 1_000_000_000_f64)) * sf
    })
   }).collect()
};

```

Reply:
Thanks, I'll try that